kotlinx-coroutines-core/common/src/channels/ArrayChannel.kt

```diff
@@ -123,6 +130,16 @@ internal open class ArrayChannel<E>(
         return receive!!.offerResult
     }
 
+    // Guarded by lock
+    private fun ensureCapacity(currentSize: Int) {
+        if (currentSize == buffer.size) {
+            val newSize = min(buffer.size * 2, capacity)
+            val newBuffer = arrayOfNulls<Any?>(newSize)
+            buffer.copyInto(newBuffer)
+            buffer = newBuffer
```

Comment:
This does not work properly when buffer is current wrapped around (`head != 0`) as the `newBuffer` will not contain a contiguous segment of values. The easiest way to fix it is to copy all the values from `buffer` into `newBuffer` using a loop into zero offset and reset head to zero like this:

```
            for (i in 0 until currentSize)
                newBuffer[i] = buffer[(head + i) % buffer.size]
            head = 0
```

Reply:
It is possible to split the loop and avoid the modulo operation.

```kotlin
val firstSliceSize = minOf(head + currentSize, buffer.size) - head
for (i in 0 until firstSliceSize) {
    newBuffer[i] = buffer[head + i]
}
for (i in 0 until head + currentSize - buffer.size) {
    newBuffer[firstSliceSize + i] = buffer[i]
}
```

Reply:
@fvasco At this point, where you have performed the mental gymnastics of getting rid of module and splitting, you can make it even better by replacing each loop with one call of `copyInto`.